### PR TITLE
Tolerant variable reference finder

### PR DESCRIPTION
**Pull request by BladeMF**
Implementing https://github.com/phpactor/phpactor/issues/1164

### CODE REVIEW

File: tests/Unit/TolerantVariableReferenceFinderTest.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            $uri,
+        ];
+
+        yield 'scope: inside class method: inside anonumous class method' => [
+            '<?php '.
+                'class C1 { function M1() { '.
+                '$var = 1;'.
+                '$c = new class { function IM() { <sr>$v<>ar = 1; } } '.
+                ' } }',
+            $uri,
```
Comment (dantleech) [inline tests/Unit/TolerantVariableReferenceFinderTest.php]: looks like we can remove the `$uri` parameter from the test as it doesn't vary?

File: tests/Unit/TolerantVariableReferenceFinderTest.php [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                    $selectionOffset = $offset;
+                } elseif ($result == "<sr>") {
+                    $referenceLocations[] = PotentialLocation::surely(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                } elseif ($result == "<mr>") {
+                    $referenceLocations[] = PotentialLocation::maybe(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                } else {
```
Comment (dantleech) [inline tests/Unit/TolerantVariableReferenceFinderTest.php]: I would not use `else` (ever almost), instead you can use `continue`:

```
if ($result === '<>) {
   // do something
   continue; // we're done
}

if ($result == '...') {
  // ...
}

// ...

throw new Exception('...');
```

you can also reducing the nesting here by removing the outer else:

```
if (!is_array($_)) {
    return [ ... ];
}

// rest of the method is not unnecessarily indented
```

File: tests/Unit/TolerantVariableReferenceFinderTest.php [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            $newSource = "";
+            $offset = 0;
+            foreach ($results as $result) {
+                if ($result == "<>") {
+                    $selectionOffset = $offset;
+                } elseif ($result == "<sr>") {
+                    $referenceLocations[] = PotentialLocation::surely(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                } elseif ($result == "<mr>") {
```
Comment (dantleech) [inline tests/Unit/TolerantVariableReferenceFinderTest.php]: I don't think this finder needs "maybe"? `maybe` would apply if we have a method `$foo->bar()` but we don't know the type of `$foo` so if we search for `Foo->bar()` it's possible that it belongs to `Foo`, but it could be an identically named method.

I think in this case we can always be 100% sure that the variable is what we're looking for?

**Comment by dantleech** [comment]
Looks good :+1: 

For testing typically I've been lazy, and would just use `[$source, $offset] = ExtractOffset::fromSource('<?php $a<>sd;')` and the assert the count of the number of references.

Happy to include this more comprehensive test though -- maybe it could be generalized in the future.

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 3 above)
Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: Just thought that code can be generalised in the future, so I included it. I never use it.

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 2 above)
Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: I choose them by gut feeling most of the time - whatever seems more readable in the current code, but I am happy to do whatever you like! Gonna fix it.

Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: How do you like that version:
```php
            foreach ($results as $result) {
                if ($result == "<>") {
                    $selectionOffset = $offset;
                    continue;
                }
                
                if ($result == "<sr>") {
                    $referenceLocations[] = PotentialLocation::surely(
                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
                    );
                    continue;
                } 
                if ($result == "<mr>") {
                    $referenceLocations[] = PotentialLocation::maybe(
                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
                    );
                    continue;
                }
                
                $newSource .= $result;
                $offset += mb_strlen($result);
            }
```

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 3 above)
Comment (dantleech) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: better to do as little as possible

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 1 above)
Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: I would seem so. Felt bad to hardcode it in the test method as it is "data" and it is technically "provided".

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 2 above)
Comment (dantleech) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: yep perfect :+1:

**Comment by BladeMF** [comment]
> Looks good 👍
> 
> For testing typically I've been lazy, and would just use `[$source, $offset] = ExtractOffset::fromSource('<?php $a<>sd;')` and the assert the count of the number of references.
> 
> Happy to include this more comprehensive test though -- maybe it could be generalized in the future.

I really liked your approach with `ExtractOffset` - it shows in a really visual way what you are testing - and expanded it to include everything _I was testing_. In the first version I described ranges by hand, which was prone to error. 

Otherwise, I am pretty detailed when it comes to testing currently, not sure how that will change in the future.

File: .gitignore [hunk 4]
```diff
@@ -1,4 +1,5 @@
 /vendor
+/.vscode
```
Comment (camilledejoye) [inline .gitignore:2]: Not the first time I see this was added recently.
IMO the project should not ignore files generated by external tools.

You should instead either ignore them globally with `~/.gitignore` or on a per project basis with `.git/info/exclude`.

File: tests/Unit/TolerantVariableReferenceFinderTest.php [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    {
+        list($source, $selectionOffset, $expectedReferences) = $this->offsetsFromSource($source, $uri);
+        $document = TextDocumentBuilder::create($source)
+            ->uri($uri)
+            ->language("php")
+            ->build();
+        
+        $finder = new TolerantVariableReferenceFinder(new Parser());
+        $actualReferences = iterator_to_array($finder->findReferences($document, ByteOffset::fromInt($selectionOffset)), false);
+        // dump($expectedReferences, $actualReferences);
```
Comment (camilledejoye) [inline tests/Unit/TolerantVariableReferenceFinderTest.php]: To remove before the merge ;)

File: composer.json [hunk 6]
```diff
@@ -21,7 +21,8 @@
         "phpactor/test-utils": "^1.0.2",
         "phpstan/phpstan": "~0.12.0",
         "phpunit/phpunit": "^9.0",
-        "phpspec/prophecy-phpunit": "^2.0"
+        "phpspec/prophecy-phpunit": "^2.0",
+        "symfony/var-dumper": "^5.2"
```
Comment (camilledejoye) [inline composer.json:26]: It's an unnecessary dependency, useful during debug but not related to the project.
I love it too, it's not an ideal solution but to avoid this issue I ended up installing it globally and configuring it on PHP directly: https://www.php.net/manual/en/ini.core.php#ini.auto-prepend-file
Again, not ideal but an interesting advantage was that you can dump with `php -a` :)

File: lib/TolerantVariableReferenceFinder.php [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    /**
+     * {@inheritDoc}
+     */
+    public function findReferences(TextDocument $document, ByteOffset $byteOffset): Generator
+    {
+        $this->foundDefinition = false;
+        $sourceNode = $this->sourceNode($document->__toString());
+        $variable = $this->variableNodeFromSource($sourceNode, $byteOffset->toInt());
+        if ($variable === null) {
+            yield from [];
```
Comment (camilledejoye) [inline lib/TolerantVariableReferenceFinder.php]: This is unnecessary if I'm not mistaken

File: lib/TolerantVariableReferenceFinder.php [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            }
+            
+            if (
+                $this->isPotentialReferenceNode($node)
+                && $name == $this->variableName($node)
+            ) {
+                if (!$this->foundDefinition) {
+                    // We artificially skip the first result
+                    // assuming it's the definition
+                    $this->foundDefinition = true;
```
Comment (camilledejoye) [inline lib/TolerantVariableReferenceFinder.php]: What is the motivation here ?
I would like to have the definition as a result too personally, especially since in PHP people tends to not declare their variables until they need to affect a value to it.
But it might be only me ;)

If it still make sense to you to ignore it, it might be better to remove the attribute `foundDefinition` from the class and make it a parameter of the method.
This notion should not get out of this method and currently it forces you to know the behavior in order to be aware that you need to set it to `true` before calling this method for the first time.
It's kind of licking everywhere I mean, not sure my explanation is that good though...

File: .gitignore (hunk 4 above)
Comment (BladeMF) [reply .gitignore:2]: There are a lot of personal preferences there. I'd rather not have somebody else's settings. Probably exclude the settings file only, then (`.vscode/settings.json`).

File: lib/TolerantVariableReferenceFinder.php (hunk 7 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: Which part?

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: The motivation was that all other reference finders did not return the definition. Probably because the definition is not a reference, not sure. The definition locator finds the first occurrence of variables as well. If I left that here then it is that location is seen twice in "Find all references".

File: composer.json (hunk 6 above)
Comment (BladeMF) [reply composer.json:26]: I get your point, but so is PHPUnit. And, you could have a versioning nightmare on your hands with Symfony projects.

Comment (dantleech) [reply composer.json:26]: I was thinking about adding this to all phpactor packages in `required-dev`, prepending it is a good solution though - although it does leave open the potential for conflicts.

Comment (BladeMF) [reply composer.json:26]: I have multiple Symfony projects in different versions, please guys.
I wonder if it could be `composer install`-ed without being in composer.json?

File: .gitignore (hunk 4 above)
Comment (dantleech) [reply .gitignore:2]: I'm fine with adding this here, it's not very intrusive and `.vscode` would probably be the biggest userbase for the language server.

Comment (BladeMF) [reply .gitignore:2]: This doesn't concern the userbase, but the contributors. The problem with this file that it contains absolute paths and things local to the developer. I can't see anybody sharing my entries there. We should probably just exclude the file, if I have to be minimalistic. See mine from another project:
```json
{
	"editor.defaultFormatter": "esbenp.prettier-vscode",
	
	"editor.insertSpaces": false,
	"editor.tabSize": 2,

	"[php]": {
		"files.eol": "\n"
	},

	"[yaml]": {
		"editor.insertSpaces": true,
		"editor.tabSize": 2,
		"editor.autoIndent": "advanced",
		"files.eol": "\n"
	},
	
	"sonarlint.ls.javaHome": "/usr/lib/jvm/java-8-openjdk-amd64/jre",
	"scss.showErrors": true,

	"files.associations": {
		"*.twig": "html"
	},

	"better-phpunit.phpunitBinary": "php7.4 bin/phpunit",

	// "phpactor.trace.server": "verbose",
	
	"phpactor.config":{
		"language_server_completion.trim_leading_dollar": true,
		"logging.level": "debug",
		"logging.path": "/root/phpactor-session.log",
		"logging.enabled": true,
		"indexer.stub_paths": [
			"/var/www/mars-ipm/app/bin/.phpunit/phpunit"
		],
		"code_transform.class_new.variants": {
			"Interface": "interface",
			"Exception": "exception",
			"Test": "test_case",
			"Symfony entity": "symfony_entity",
			"Symfony use case": "symfony_use_case",
			"Symfony use case interface": "symfony_use_case_interface",
			"Symfony kernel test case": "symfony_kernel_test_case",
		}
	},
	"sqltools.connections": [
		{
			"previewLimit": 50,
			"driver": "SQLite",
			"name": "Tests",
			"database": "app/tests/tests.sqlite"
		}
	],
	"sqltools.useNodeRuntime": true
}
```

Comment (dantleech) [reply .gitignore:2]: yes, i just meant to imply that the more of a userbase we have the more contributors we are likely to have. Saying that this is the type of thing that I ignore globally.

Comment (camilledejoye) [reply .gitignore:2]: I'm not sure to see the problem with using `.git/info/exclude` but it's not a big deal, @BladeMF you can resolve (I can't)

File: composer.json (hunk 6 above)
Comment (camilledejoye) [reply composer.json:26]: https://github.com/franzliedke/studio was doing this, but I don't know if they updated to composer 2

I'm well aware of the issues that might bring a global version, I deal with it every day :)
My goal here is not to be annoying, as painful as it is to locally deal with versioning issue it's even worse with phpactor because of the number of packages.
It could be a pain later to have to deal with Symfony dependencies since we have none yet in this package.

Since @dantleech was already thinking about adding it globally it's a non issue so you can resolve ;)
I didn't require any changes with this review, it's only there to gave feedback and have the opportunity to share.

File: .gitignore (hunk 4 above)
Comment (BladeMF) [reply .gitignore:2]: @camilledejoye, if I am reading this right, this setting will be local for me. The problem with that approach is if another guy who uses VS Code contributes and then his settings end up in the repo which defeats the purpose, because they will interfere with mine.

File: lib/TolerantVariableReferenceFinder.php (hunk 7 above)
Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: The `yield from []`, you shoul'd be able to simply return.
That's how it's done in the completors if my memory is correct.

File: composer.json (hunk 6 above)
Comment (BladeMF) [reply composer.json:26]: I welcome the feedback!

File: lib/TolerantVariableReferenceFinder.php (hunk 7 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: I will try this again, but if I am not mistaken, I got an error last time.

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: OK, for me a reference is just a place where the variable is used.
I don't make a difference between declaration, assignation, usage, etc.

> The definition locator finds the first occurrence of variables as well

That's its purpose, I guess you meant that it uses the same logic: the first time we see the variable is the definition
I've nothing against this idea, since it's impossible to manually declare a variable in PHP it make perfect sense.

What do you call the "Find all references" exactly ? Which handler ?
If it's shown twice it would suggest that the logic in this feature is to separate the definition of the reference indeed.
But then when does this `TolerantVariableReferenceFinder` would be used ?

Basically I just tried to make sure that when we ask for the references of a variables we have all the results, if it's aggregated with a definition locator great then.
Like currently if I leave my cursor on a variable it hovers all the occurrences, including the declaration.

File: composer.json (hunk 6 above)
Comment (dantleech) [reply composer.json:26]: > even worse with phpactor because of the number of packages.

that's not so problematic now, i can update all [packages in a few minutes](https://github.com/phpactor/phpactor-hub)

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: > What do you call the "Find all references" exactly ? Which handler ?

`Phpactor\Extension\LanguageServerReferenceFinder\Handler\ReferencesHandler`

> But then when does this TolerantVariableReferenceFinder would be used ?

What do you mean? My point is that I just followed the way every other `ReferenceFinder` works - to not return the definition. I am fine with either way, but it must be consistent between all the `ReferenceFinder`s, I think.

>Basically I just tried to make sure that when we ask for the references of a variables, we have all the results, if it's aggregated with a definition locator great then.

I agree. I took the code form the `ReferencesHandler`, because to rename something you are basically renaming a subset of the references, but you need to check them all. There the references were combined with the defintion, so I went along. I don't mind reworking my code. I agree with your logic, but I had to be consistent with the existing code.

> Like currently if I leave my cursor on a variable it hovers all the occurrences, including the declaration.

I don't think the Hover handler uses references at all. It just highlight the text which includes the variables with the same name in other methods.

File: .gitignore (hunk 4 above)
Comment (dantleech) [reply .gitignore:2]: We wouldn't let anybody commit their `.vscode` folder though :)

Comment (BladeMF) [reply .gitignore:2]: I that case, I'll do whatever you say. But isn't that the purpose of `.gitignore` - to not let people commit stuff?

Comment (camilledejoye) [reply .gitignore:2]: For me the goal is to avoid me the pain to deal with it but I get your point.
I personally always use my global gitignore when it comes to stuff that will impact multiple project and the `.git` directory for stuff that are specific to a project.

Comment (dantleech) [reply .gitignore:2]: Yes, I've added this to all Phpactor repos already, but yes, this file should be used for project-specific things (and we shouldn't add the ignore patterns for every single IDE ever made)

Comment (BladeMF) [reply .gitignore:2]: I totally believe you have good reasons for doing so, but for me the `.gitignore` file is sort of "The rules of the repository" and it goes with the repository for everyone to follow. I could be missing something and I would be grateful if you explained if I did.

Comment (BladeMF) [reply .gitignore:2]: > and we shouldn't add the ignore patterns for every single IDE ever made

I get your point - people should be well-mannered enough to not commit their stuff. I will do whatever you instruct me to.

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: Thanks for the handler, make more sense to me know, depending on the context we will add the definition or not.
So you're right to keep it like that, I was worried that if we tried to implement the rename with only the finder we would endup not renaming the definition.

But, for me who likes to see the definition it will work because I will configure the context accordingly.
For the rename we will have to be aware that will also need to retrieve the definition.

What about the suggestion of removing the class attribute and pass it as arguments to the method ?

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 1 above)
Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: Agreed.

File: lib/TolerantVariableReferenceFinder.php (hunk 7 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: You were correct.

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: Actually I liked your suggestion and just tried it. It didn't work. I didn't have enough time to read, but would seem that you can't change a variable within the `foreach`, because the yields seem to be executed in some different order or something. The problem was the recursive call - when the definition was found, the `true` value was used only for _that_ cycle, but not the next cycles, if that makes sense.
If you think I implemented it wrong, I can check in the non-working code so you can take a look.

Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: Something like:
```php
    /**
     * @return Generator<PotentialLocation>
     */
    private function find(Node $scopeNode, string $name, string $uri, bool $foundDefinition = false): Generator
    {
        /** @var Node $node */
        foreach ($scopeNode->getChildNodes() as $node) {
            if ($node instanceof AnonymousFunctionCreationExpression && !$this->nameExistsInUseClause($name, $node)) {
                continue;
            }
            
            if (
                $this->isPotentialReferenceNode($node)
                && $name == $this->variableName($node)
            ) {
                if (!$foundDefinition) {
                    // We artificially skip the first result
                    // assuming it's the definition
                    $foundDefinition = true;
                } else {
                    yield PotentialLocation::surely(
                        Location::fromPathAndOffset($uri, $node->getStart())
                    );
                }
            } else {
                yield from $this->find($node, $name, $uri, $foundDefinition);
            }
        }
    }
```

Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: That is almost exactly what I tried. If I had to guess, at the time `yield from $this->find($node, $name, $uri, $foundDefinition);` was executed, the value of `$foundDefinition` is already different. When I did some dumping I saw that after the iteration where `$foundDefinition` became true, it was again false for the next iteration. Not sure what am I doing wrong.
Will try again and post the results.

Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: Boy am I dumb. This doesn't work because this is a recursion rather than a flat `foreach`-ing. If a call stack set the argument to `true`, we want all other branches to know about it. Look at this call stack:
```
.^ "Found: "
^ "Found: "
^ "  Found: "
^ "    Found: "
^ "    Found: 1" <------ Here it finds the definition
^ "Found: " <---- We now want this to know about it as well
^ "  Found: "
^ "    Found: "
^ "    Found: "
^ "      Found: "
^ "      Found: 1"
```
This means we should pass it by reference, which I am not terribly fond of.

File: phpstan.neon [hunk 9]
```diff
@@ -1,7 +1,8 @@
 includes:
     - phpstan-baseline.neon
 parameters:
+    level: 7
     inferPrivatePropertyTypeFromConstructor: true
+    checkMissingIterableValueType: false
```
Comment (dantleech) [inline phpstan.neon]: If possible I'd prefer to keep this `true`

**Comment by dantleech** [comment]
How's this going? To what extent does it compare to https://github.com/phpactor/code-transform/blob/7419a24f702826baad09c0c118560771d83c3276/lib/Adapter/TolerantParser/Refactor/TolerantRenameVariable.php in regards to finding references?

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: I'll fix it.

**Comment by BladeMF** [comment]
> How's this going? To what extent does it compare to https://github.com/phpactor/code-transform/blob/7419a24f702826baad09c0c118560771d83c3276/lib/Adapter/TolerantParser/Refactor/TolerantRenameVariable.php in regards to finding references?

I did start with this code in the first place. Then expanded on it as needed. Except for the issue @camilledejoye pointed out with the instance variable I think we're on the same page. I'd say whenever he's happy we can merge.

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: Interestingly enough, these are the reported errors:
```
------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   WorsePlainTextClassDefinitionLocator.php                                                                                                                         
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  27     Property Phpactor\WorseReferenceFinder\WorsePlainTextClassDefinitionLocator::$breakingChars type has no value type specified in iterable type array.             
         💡 Consider adding something like array<Foo> to the PHPDoc.                                                                                                       
         You can turn off this check by setting checkMissingIterableValueType: false in your phpstan.neon.                                                                
  34     Method Phpactor\WorseReferenceFinder\WorsePlainTextClassDefinitionLocator::__construct() has parameter $breakingChars with no value type specified in iterable   
         type array.                                                                                                                                                      
         💡 Consider adding something like array<Foo> to the PHPDoc.                                                                                                       
         You can turn off this check by setting checkMissingIterableValueType: false in your phpstan.neon.                                                                
  143    Method Phpactor\WorseReferenceFinder\WorsePlainTextClassDefinitionLocator::resolveImportTable() return type has no value type specified in iterable type array.  
         💡 Consider adding something like array<Foo> to the PHPDoc.                                                                                                       
         You can turn off this check by setting checkMissingIterableValueType: false in your phpstan.neon.                                                                
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------- 
```
Not sure why _I_ changed this setting...

Comment (BladeMF) [reply phpstan.neon]: I'd appreciate suggestions. My laptop will be in the repair shop tomorrow, but after that I can make changes.

Comment (dantleech) [reply phpstan.neon]: I added phpstan at level 7 to all repos recently and generated a baseline for all of them, maybe this was created before that change. I think if you rebase they will go away, otherwise I think there are only a few failures as above so should be easy to eradicate them and remove the baseline.

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: Using reference is OK for me, it's not a bad thing as long as it is justified.
When you think about it, before you're using a class attribute which have a global scope. This is not really different, if you're uncomfortable with it because you think it's unclear you can always document it in the phpdoc to explain it a bit more.

Comment (dantleech) [reply lib/TolerantVariableReferenceFinder.php]: Instead of doing that, why not just discard the first result? This would also free this method of that responsiblity

Comment (camilledejoye) [reply lib/TolerantVariableReferenceFinder.php]: Indeed, I didn't find any indication in the documentation that would suggest `yield from` does not keep the ordering.
The find method could then be used to return all references, included the definition.

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: I'll try rebasing

File: lib/TolerantVariableReferenceFinder.php (hunk 8 above)
Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: Who should discard it? And how?

Comment (dantleech) [reply lib/TolerantVariableReferenceFinder.php]: Delegate to anoter method:

```
$generator = $this->doFindReference($blah);
// discard first entry
$generator->next();
yield from $generator;
```

or similar, probably a better way :smile: 

in the long run though, if we only want to exclude the definition, we should explicitly detect the definition (but for now it's fine)

Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: > Using reference is OK for me, it's not a bad thing as long as it is justified.
> When you think about it, before you're using a class attribute which have a global scope. This is not really different, if you're uncomfortable with it because you think it's unclear you can always document it in the phpdoc to explain it a bit more.

I think it would make the code less readable, yes. And it pollutes all the calls with the initialisation of the variable, especially the first one where it is not relevant, because it is internal to the method. It is, in fact, mimicking an instance variable for a smaller scope.

Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: > $generator = $this->doFindReference($blah);

I actually like that. Making the change.

Comment (BladeMF) [reply lib/TolerantVariableReferenceFinder.php]: Pushed.

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: We now have several failing tests. Can you take a look?

Comment (dantleech) [reply phpstan.neon]: hmm looks good to me?

Comment (BladeMF) [reply phpstan.neon]: composer integrate failed on my machine. Probably out of date dependencies?

File: tests/Unit/TolerantVariableReferenceFinderTest.php [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+                    $referenceLocations[] = PotentialLocation::maybe(
+                        new Location($textDocumentUri, ByteOffset::fromInt($offset))
+                    );
+                    continue;
+                }
+                
+                $newSource .= $result;
+                $offset += mb_strlen($result);
+            }
+        } else {
```
Comment (dantleech) [inline tests/Unit/TolerantVariableReferenceFinderTest.php]: can we invert this and remove the dangling else:

```
if (!is_array($blah)) {
    throw new Exception(...');
}

// do the stuff here
```

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: If you point me to which I should update, it would save me a ton of time tomorrow.

**Comment by dantleech** [comment]
LGTM?

File: phpstan.neon (hunk 9 above)
Comment (dantleech) [reply phpstan.neon]: what are the failures?

i mean, updating phpstan should do the trick

Comment (dantleech) [reply phpstan.neon]: also try clearing the result cache `phpstan clear-result-cache`

File: tests/Unit/TolerantVariableReferenceFinderTest.php (hunk 10 above)
Comment (BladeMF) [reply tests/Unit/TolerantVariableReferenceFinderTest.php]: Sure. I actually rather like this approach, but you see it faster than I do.

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]:
```
1) Phpactor\WorseReferenceFinder\Tests\Unit\WorsePlainTextDefinitionLocatorTest::testGotoWord with data set "xml attribute" ('<element class="Foob<>ar">', 'Foobar.php')
Phpactor\ReferenceFinder\Exception\CouldNotLocateDefinition: Word ""Foobar"" could not be resolved to a class

/root/phpactor/worse-reference-finder/lib/WorsePlainTextClassDefinitionLocator.php:56
/root/phpactor/worse-reference-finder/tests/DefinitionLocatorTestCase.php:21
/root/phpactor/worse-reference-finder/tests/Unit/WorsePlainTextDefinitionLocatorTest.php:18

Caused by
Phpactor\WorseReflection\Core\Exception\SourceNotFound: Could not find source with ""Foobar""

/root/phpactor/worse-reflection/lib/Core/SourceCodeLocator/ChainSourceLocator.php:58
/root/phpactor/worse-reflection/lib/Core/Reflector/CoreReflector.php:126
/root/phpactor/worse-reflection/lib/Core/Reflector/CompositeReflector.php:73
/root/phpactor/worse-reference-finder/lib/WorsePlainTextClassDefinitionLocator.php:54
/root/phpactor/worse-reference-finder/tests/DefinitionLocatorTestCase.php:21
/root/phpactor/worse-reference-finder/tests/Unit/WorsePlainTextDefinitionLocatorTest.php:18

Caused by
Phpactor\WorseReflection\Core\Exception\SourceNotFound: Could not find source for ""Foobar"" in stub directory "/root/phpactor/worse-reference-finder/tests/Workspace"

/root/phpactor/worse-reflection/lib/Core/SourceCodeLocator/StubSourceLocator.php:43
/root/phpactor/worse-reflection/lib/Core/SourceCodeLocator/ChainSourceLocator.php:42
/root/phpactor/worse-reflection/lib/Core/Reflector/CoreReflector.php:126
/root/phpactor/worse-reflection/lib/Core/Reflector/CompositeReflector.php:73
/root/phpactor/worse-reference-finder/lib/WorsePlainTextClassDefinitionLocator.php:54
/root/phpactor/worse-reference-finder/tests/DefinitionLocatorTestCase.php:21
/root/phpactor/worse-reference-finder/tests/Unit/WorsePlainTextDefinitionLocatorTest.php:18
```

**Comment by BladeMF** [comment]
> LGTM?

=?

File: .gitignore (hunk 4 above)
Comment (BladeMF) [reply .gitignore:2]: What should I do here?

File: phpstan.neon (hunk 9 above)
Comment (BladeMF) [reply phpstan.neon]: I updated everything. It's gone.

Comment (dantleech) [reply phpstan.neon]: ah. it required the latest `phpactor/text-document`

**Comment by dantleech** [comment]
LGTM? = let's go to merge? / shall we merge it?

File: .gitignore (hunk 4 above)
Comment (dantleech) [reply .gitignore:2]: well, I've added it to all the repos (or most of them, not sure it worked properly), so leave it. But as a policy I think we should not have IDE specific patterns here, so I should probably remove them all again :roll_eyes:
